File: src/com/wang/Game2dEngine/action/ScaleHeightToAction.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action;

import com.wang.Game2dEngine.sprite.Sprite;

import static java.lang.Math.abs;

/**
 * @author ricolwang
 */
public class ScaleHeightToAction extends ScaleAction
{

    double scaleHeightTo;
    float scaleHeightToDuration;
    ScaleHeightByAction theScaleHeightByAction;

    public ScaleHeightToAction(Sprite theSprite)
    {
        this.setSprite(theSprite);
        this.scaleHeightTo = this.theSprite.getHeight();
        this.scaleHeightToDuration = 0;
    }

    public void ScaleHeightTo(double height, float duration)
    {
        this.scaleHeightTo = height;
        this.scaleHeightToDuration = abs(duration * 1000);
    }

    @Override
    public void clearSprite()
    {
        this.theScaleHeightByAction.clearSprite();
        this.theScaleHeightByAction = null;
        this.theSprite = null;
    }

    @Override
    public void perform(double runningTime)
    {
        if (bComplete)
        {
            return;
        }

        if (theScaleHeightByAction == null)
        {
            theScaleHeightByAction = new ScaleHeightByAction();
            double height = this.theSprite.getHeight();
            theScaleHeightByAction.scaleHeightBy(scaleHeightTo - height, this.scaleHeightToDuration / 1000.0f);
            theScaleHeightByAction.setSprite(theSprite);
        }

        theScaleHeightByAction.perform(runningTime);
        if (theScaleHeightByAction.bComplete)
        {
            bComplete = true;
        }
    }
}

File: src/com/wang/Game2dEngine/action/ScaleHeightToAction.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action

import com.wang.Game2dEngine.sprite.Sprite

/**
 * @author ricolwang
 */
class ScaleHeightToAction(theSprite: Sprite?) : ScaleAction()
{
    var scaleHeightTo: Double
    var scaleHeightToDuration: Float
    var theScaleHeightByAction: ScaleHeightByAction? = null
    fun ScaleHeightTo(height: Double, duration: Float)
    {
        scaleHeightTo = height
        scaleHeightToDuration = Math.abs(duration * 1000)
    }

    override fun clearSprite()
    {
        theScaleHeightByAction!!.clearSprite()
        theScaleHeightByAction = null
        theSprite = null
    }

    override fun perform(runningTime: Double)
    {
        if (bComplete)
        {
            return
        }
        if (theScaleHeightByAction == null)
        {
            theScaleHeightByAction = ScaleHeightByAction()
            val height = theSprite!!.height
            theScaleHeightByAction!!.scaleHeightBy(scaleHeightTo - height, scaleHeightToDuration / 1000.0f)
            theScaleHeightByAction!!.setSprite(theSprite)
        }
        theScaleHeightByAction!!.perform(runningTime)
        if (theScaleHeightByAction!!.bComplete)
        {
            bComplete = true
        }
    }

    init
    {
        setSprite(theSprite)
        scaleHeightTo = this.theSprite!!.height
        scaleHeightToDuration = 0f
    }
}